### DISCUSSION TITLE
Supabase realtime chat app

### DISCUSSION THREAD

**Discussion by iampopal** [supabase/supabase#8987] [Questions] [answered]
We want to create a WhatsApp like chat app

we have the following tables

rooms
room_participants(room_id, user_id)
messages(id,content, sender, room_id)

we want to get all messages of rooms which which are send to current logged in user

**Comment by iampopal** [comment]
When using supabase.from('messages').stream('id')

We want to listen to all messages of which are of this user

**Comment by GaryAustin1** [comment, marked as answer]
Realtime only supports a .eq filter at this time (I'm not sure about streams) so you would need a separate connection for each room if you just have your current tables.  A .in filter would be ideal.

What I did was create a separate room status table with RLS allowing only user to see the status of rooms they are in.
Then I have an insert trigger function on messages that updates last message info in that room's status row.
Realtime monitors all rows in the status table, but because of RLS you only see changes in the rooms you are in.

Chat for my application is an extra feature and not the main focus.  For a low volume of messages, I feel comfortable using this trigger method with a single realtime connection is better than a connection for each room per user, performance and data bandwidth wise.

**Reply by GaryAustin1** [reply]
You should not start new answers to follow up....

You can not do joins in the realtime subscription.

If you know you need to do a join each time, you could just update with the message id in the status table and then when you get the realtime event on the client use that id to select the message and join anything you need in a database request.  Even if you were to have a subscription on every room in the message table you would still need to do a database request to get join information, so not really an extra overhead.

Edit:   I guess you might be able to collect your join info in the trigger function and include all of that in the status row update to avoid a call back to the database. My case I don't need to join user info.

**Reply by iampopal** [reply]
Thank you for your great answer, 

yes, I am thinking to do create a room_status table that will insert or update recent records of room, and the client side will subscribe to the room_status table

**Reply by iampopal** [reply]
Any other suggestions for us to keep while creating a real-time chat app? best practices to make app have great performance.
